### DISCUSSION TITLE
Sparsemax

### DISCUSSION THREAD

**Discussion by wangzcl** [aleixalcacer/archetypes#67] [Ideas]
I tried the idea of using softmax to implement a Pytorch version of AA and avoid manual gradient. It turns out that it is not as accurate as the simplex PGD version. I think the problem is that softmax disfavors sparse outputs (gradient vanishes when outputs saturate).

There are some softmax alternatives that will definitely solve the problem, for example, sparsemax  ([deep-spin/entmax](https://github.com/deep-spin/entmax)) which is exactly a same simplex projection operator using an older algorithm. One potential problem is that some papers say that it is much slower, though I've never tried. It obviously involves a lot of buffer allocation, which might be alleviated using `torch.compile`.
